packages/yew-router/src/utils.rs

```diff
@@ -55,7 +55,7 @@ pub fn compose_path(pathname: &str, query: &str) -> Option<String> {
         })
 }
 
-#[cfg(not(target_arch = "wasm32"))]
+#[cfg(any(not(target_arch = "wasm32"), target_os = "wasi"))]
```

Comment:
Given the amount of work required to make this change, I don't think it's justified to make this change in this PR, although I fully support this idea.

Maybe the reason I gave is a bit embarrassing. In order to support this feature, I have to go back and improve the implementation of `gloo` to ensure that it can simulate a batch of virtual interfaces when running on non-WASM targets. After all this work is completed, I am worried that you will have no one to review it for at least a few months...

This proposal can be made based on a new PR. And of course, this must be done after this PR is processed. :P

Reply:
I should have been more clear that I mostly have taken the space to write down my thoughts about the situation as a whole, in case multiple different platforms come along on which people want to run SSR on. Not a blocker for this PR.

Reply:
> I should have been more clear that I mostly have taken the space to write down my thoughts about the situation as a whole, in case multiple different platforms come along on which people want to run SSR on. Not a blocker for this PR.

There is a small problem. `csr` and `ssr` can be enabled at the same time. Developers can of course distinguish them by developing in different crates, using `csr`/`hydration` and `ssr` respectively. However, under the same workspace, the Rust compiler will enable all these features at the same time during analysis.

I can certainly make improvements based on the solution you mentioned. The only problem is that we still have to use platform checks to distinguish compilation targets, because both `csr` and `ssr` are enabled at this time. If an alternative facility is used for `gloo` at this time, it may cause conflicts.

So I still have doubts about this. If adding features checks is the same as not adding them, how can I be sure that my changes are correct?

Reply:
> There is a small problem. csr and ssr can be enabled at the same time. Developers can of course distinguish them by developing in different crates, using csr/hydration and ssr respectively.

Yes, correct. And the initial approach and the current state of affairs is that the codepath is decided based on the _platform_ not on the render facility used. So far, that was a usable approach. I believe by targeting `wasi`, this assumption breaks down though (and to be clear it has been on shaky grounds). The `wasm-wasi` target could possibly be used either in the browser or in a backend. For this PR, I think it is fine to say that we still only support `wasm-unknown-unknown` as a browser environment, and anything else is a server backend for us. But in the not-too-distant future, we should support `wasi` also on the client side (since it adds POSIX compat) and any `wasm` on the server side (to easily load in azure, aws etc when you already have a js server).

In this case, we should try the approach outlined above, where we have codepaths for csr and ssr but they can both be implemented and live next to each other for a shared set of platforms (and mocked out for others), but are chosen based on the code path used to render the app.

> I can certainly make improvements based on the solution you mentioned.

If you can see a way of easily building in the assumption of "On this platforms, we are rendering on a server and on these other ones, it's a client" in a central place instead of repeating `#[cfg(all(target_ ... ))]` all over the place, that would be cool, but it will take some work that I think is not yet necessary if we replace it later anyway.

Reply:
Ah I can see you made some changes already trying to address this, I will try to review this evening or tomorrow.